add new stream uploader

brendanashworth opened the pull request:
This PR adds a new stream uploader, `.streamObject()`, which allows for easy uploading of a stream to s3.

This also ports `.fPutObject()` and `.putObject()` to use `.streamObject()` under the hood.

**NOTE**: this contains a breaking change. Before this, uploading a stream of size 1GB (for example) but declaring it as 2GB would leave the multipart upload open on the server. This PR now makes it close the multipart upload once the stream ends. The tests have been updated/removed for that.

harshavardhana commented:
This is actually an important breaking API change.  Size is provided for a requirement to calculate optimal part size.  Also to seamlessly resume a previously uploaded object.

harshavardhana commented:
Will ping you lets discuss this further

harshavardhana commented:
Will review @brendanashworth tonight.

krishnasrinivas commented:
@brendanashworth  there is one important requirement for fPutObject, it should not hold anything in memory. If you look at current putObject() it has no choice but to hold it up in the memory because once it reads from the stream for md5sum and sha256 calculation it can't just drop it as the read stream is not seekable to seek and read it again. So for large objects this won't be fun for a 5TB upload the partsize will be around 500MB and nodejs might not even work properly. But if you look at current fPutObject() implementation it does not hold anything in memory at all so it will work smoothly regardless of the objectsize/partsize. fPutObject is written the way it is as nodejs streams are not seekable unlike how it is in other languages like Go

krishnasrinivas commented:
@brendanashworth another requirement is for small objects (less than `minimumPartSize`) we should not upload as multipart and instead do direct PUT

harshavardhana commented:
@brendanashworth @krishnasrinivas sums up pretty much what i wanted to say.

harshavardhana commented:
Let's talk again on this tomorrow.

brendanashworth commented:
I talked with @harshavardhana and it seems like the following is what we'd like to do with this:

- to check out memory usage first before moving ahead
- drop `fPutObject`, as a different approach to implementation is better, reading the file in chunks and uploading each chunk individually instead of streaming it
- drop the new `streamObject` API and keep the new functionality only in `putObject`

@krishnasrinivas thanks for your comments. I did implement the single-request PUT in the latest version, but your `fPutObject` analysis is spot-on.

Regarding issue you brought up about `putObject` buffering too much memory, Harsha and I talked about this a little bit — to fix it, I created [`block-stream-file`](https://github.com/brendanashworth/block-stream-file), which is a drop-in replacement for [`BlockStream2`](https://github.com/minio/minio-js/blob/588f1399fd4743f9a0f515333be664330bc57127/src/main/minio.js#L1036). I'm going to do some memory profiling before I am *sure* the replacement would fix the memory usage, but when I introduce it into this PR we should see a rather dramatic improvement in overall memory usage.

I'll still work on this PR, I'll try and finish it up soon so it doesn't get stale in regards to the codebase.

harshavardhana commented:
> drop fPutObject, as a different approach to implementation is better, reading the file in chunks and uploading each chunk individually instead of streaming it

We need to keep 'fPutObject' just that we have to use NodeJS readOffset functions.

brendanashworth commented:
> We need to keep 'fPutObject'

Yes, I just meant drop it from the pull request — sorry!

krishnasrinivas commented:
> Regarding issue you brought up about putObject buffering too much memory, Harsha and I talked about this a little bit — to fix it, I created block-stream-file

we had discussed using tmp files instead of keeping it in memory, but AB did not like it as it (I think because of the side effects it brings in the system like extra maintenance work of deleting tmp files, ramdisk based tmpfs eating memory etc) Also for majority of the cases - till 50GB the chunk size is 5MB which is fine.

Current putObject() is fine memory wise because it keeps just 5MB in memory, by memusage concern in https://github.com/minio/minio-js/issues/428 what I was curious was  to find is if streams work as expected without buffering up unnecessarily in memory. Unless there is a bug in our code this should not happen for even files up to 50G.

> We need to keep 'fPutObject' just that we have to use NodeJS readOffset functions.

you mean `fs.read(fd, buffer, offset, length, position, callback)` ? I had explored this API when I wrote fPutObject but found a better API for our purpose - fs.createReadStream() that I used i.e it gives a stream directly with the "seek" effects as you can specify start and end offsets.

@brendanashworth if you can list the things you are trying to achieve in this patch that will help me to evaluate/suggest. I had tried to unify code for putObject and fPutObject but it got complicated as the two APIs had different requirements. putObject needed to work on streams and fPutObject needed to be memory efficient. Added to this nodejs streams does not support seek() call. So keeping the code separate was the cleanest approach. But there will always be scope to do things better. So let me know what problem you are trying to solve which will help me give you ideas.

brendanashworth commented:
Forgive me for the prolonged absence 😅  Been a bit busy.

@krishnasrinivas you're entirely correct. The expectations of the `fPutObject` and `putObject` APIs are different. Reconciling the two is beyond our reach for now. I've removed any changes to `fPutObject` from this pull request because of that, thank you.

The only problem I'm trying to fix in this pull request is the necessity of the **size parameter** in `putObject`. I want to make that parameter *optional*. Node.js streams by nature are limitless — most core and userland stream APIs do not have the foresight to know the length of the entire stream before it ends. In my opinion, minio-js should support the uploading of these streams, even if it isn't *performant* but because it's the only *possible* way.

As an example, if you wanted to stream an HTTP PUT upload (with no size specified by the client) through a Node.js server and into s3 with minio-js, *currently* you'd have to buffer the entire upload on your application-side. Because we require you to know the size beforehand, you need to build in your own buffering app because it's simply *impossible* to know the final upload size.

This PR's intention is to make it optional by defaulting to the largest multipart size possible. If the API user knows the size beforehand, they can still provide it to receive better performance. If not, they'd be buffering the data anyways, and now we can do it for them.

I've updated the pull request to reflect this. I hope you like it and I'm always willing to make changes 😄

harshavardhana commented:
Welcome back @brendanashworth :-) - will review the PR. Thanks.

harshavardhana commented on docs/API.md:
````diff
@@ -477,7 +477,7 @@ minioClient.fGetObject('mybucket', 'photo.jpg', '/tmp/photo.jpg', function(err)
 ```
 <a name="putObject"></a>
-### putObject(bucketName, objectName, stream, size, contentType, callback)
+### putObject(bucketName, objectName, stream, size, callback)
````
contentType is still needed when a user wishes to provide contentType . IT is a necessary input parameter.

brendanashworth replied:
Should I put it back in the documentation? Right now the parameter is still *supported*, it's just legacy optional — the new `ObjectUploader` doesn't require it. Or should I change something under the hood?

harshavardhana replied:
It is better to mention it in docs otherwise no one will know how to use it and it is consistent with our other libs as well.

brendanashworth replied:
Good idea, fixed!

harshavardhana commented on src/main/object-uploader.js:
```diff
@@ -0,0 +1,224 @@
+      return
+    }
+
+    // Continue flushing various parts if we're initiated.
```
```// Continue uploading various parts if we have initiated multipart upload. ```

harshavardhana commented on src/main/object-uploader.js:
```diff
@@ -0,0 +1,224 @@
+      return
+    }
+
+    // This is called when all of the chunks are finished uploading and the
```
``` // This is called when all of the chunks uploaded successfully, thus complete the multipart upload. ```

harshavardhana commented on src/main/object-uploader.js:
```diff
@@ -0,0 +1,224 @@
+      this.client.findUploadId(this.bucketName, this.objectName, (err, id) => {
+        if (err) return this.emit('error', err)
+
+        // If no upload ID exists, get our own.
```
``` // If no upload ID exists, initiate a new one. ```

harshavardhana commented on src/main/object-uploader.js:
```diff
@@ -0,0 +1,224 @@
+        this._transform(chunk, encoding, callback)
+      })
+
+      // Check for an old upload.
```
``` Check for an incomplete previous upload. ```

brendanashworth commented:
Thanks @harshavardhana, I've changed the comments. Could you double check that the breaking changes are desired — thanks.
